lib/stages/VisWriter.cpp

```diff
@@ -0,0 +1,379 @@
+
+#include "VisWriter.hpp"
+
+#include "Config.hpp"            // for Config
+#include "Stage.hpp"             // for Stage
+#include "StageFactory.hpp"      // for REGISTER_KOTEKAN_STAGE, StageMakerTemplate
+#include "buffer.h"              // for Buffer
+#include "bufferContainer.hpp"   // for bufferContainer
+#include "datasetManager.hpp"    // for dset_id_t, fingerprint_t
+#include "prometheusMetrics.hpp" // for Counter, MetricFamily
+#include "restServer.hpp"        // for connectionInstance
+#include "version.h"             // for get_git_commit_hash
+#include "visBuffer.hpp"         // for VisFrameView
+#include "visFile.hpp"           // for visFileBundle, visCalFileBundle
+#include "visUtil.hpp"           // for movingAverage
+
+#include <cstdint>   // for uint32_t
+#include <errno.h>   // for ENOENT, errno
+#include <future>    // for future
+#include <map>       // for map
+#include <memory>    // for shared_ptr, unique_ptr
+#include <mutex>     // for mutex
+#include <set>       // for set
+#include <stdexcept> // for runtime_error
+#include <stdio.h>   // for size_t, remove
+#include <string>    // for string, operator+
+#include <unistd.h>  // for access, F_OK
+#include <utility>   // for pair
+
+using kotekan::bufferContainer;
+using kotekan::Config;
+using kotekan::Stage;
+using kotekan::prometheus::Metrics;
+
+using kotekan::connectionInstance;
+using kotekan::HTTP_RESPONSE;
+using kotekan::restServer;
+
+REGISTER_KOTEKAN_STAGE(VisWriter);
+REGISTER_KOTEKAN_STAGE(VisCalWriter);
+
+VisWriter::VisWriter(kotekan::Config& config, const std::string& unique_name,
+                     kotekan::bufferContainer& buffer_container) :
+    BaseWriter(config, unique_name, buffer_container){};
+
+void VisWriter::main_thread() {
+
+    frameID frame_id(in_buf);
+
+    kotekan::prometheus::Gauge& write_time_metric =
+        Metrics::instance().add_gauge("kotekan_writer_write_time_seconds", unique_name);
+
+    std::unique_lock<std::mutex> acqs_lock(acqs_mutex, std::defer_lock);
+
+    while (!stop_thread) {
+
+        // Wait for the buffer to be filled with data
+        if (wait_for_full_frame(in_buf, unique_name.c_str(), frame_id) == nullptr) {
+            break;
+        }
+
+        // Get a view of the current frame
+        auto frame = VisFrameView(in_buf, frame_id);
+        write_data(frame, write_time_metric, acqs_lock);
+
+        // Mark the buffer and move on
+        mark_frame_empty(in_buf, unique_name.c_str(), frame_id++);
+
+        // Clean out any acquisitions that have been inactive long
+        close_old_acqs();
+    }
+}
```

Comment:
The main threads look super similar in both `HFBWriter` and `VisWriter`. Can you move the main thread in `BaseWriter`?

Reply:
I think you'll need to change the API of `write_data` to take `(in_buf, frame_id)`